### PR TITLE
Multipath support for OSD on PVC

### PR DESCRIPTION
**Pull request by travisn**
**Description of your changes:**
When configuring an OSD on a multipath device, ceph-volume inventory is failing to detect that the device is available, complaining about insufficient space. 
```
ceph-volume inventory --format json /dev/mapper/3600605b000ac584025e796091b840018
{"available": false, "lvs": [], "rejected_reasons": ["Insufficient space (<5GB)"], "sys_api": {}, "path": "/dev/mapper/3600605b000ac584025e796091b840018", "device_id": ""}
```
Instead of relying on c-v inventory to determine if the device is available, just call c-v raw list to determine if there is already an osd configured on the pv. Otherwise, just attempt to configure the OSD. Instead we should get a fix from c-v, but this will allow testing to see if the OSD config can get path that error while preparing the OSD.      

**Checklist:**

- [ ] **CommitLint Bot**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[test ceph]

### CODE REVIEW

**Comment by satoru-takeuchi** [comment]
This fix looks like allowing not only "mpath", but also "crypt" (or may also "linear"?). It's because sys.Crypt is already allowed in `supportDeviceType()` and this type can also bypass `ceph-volume inventory` check.

Is my understanding correct?

File: pkg/daemon/ceph/osd/daemon.go [hunk 1]
```diff
@@ -319,9 +318,26 @@ func getAvailableDevices(context *clusterd.Context, desiredDevices []DesiredDevi
...
 		// So earlier lsblk extracted the '/dev' path, hence the device.Name property
 		// device.Name can be 'xvdca', later this is formated to '/dev/xvdca'
-		isAvailable, rejectedReason, err := sys.CheckIfDeviceAvailable(context.Executor, device.RealPath, pvcBacked)
-		if err != nil {
-			return nil, errors.Wrapf(err, "failed to get device %q info", device.Name)
+		var err error
+		var isAvailable bool
+		rejectedReason := ""
+		if agent.pvcBacked {
+			block := fmt.Sprintf("/mnt/%s", agent.nodeName)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/daemon.go:325]: What about keeping `device.RealPath`? (use `/dev/` then).

File: pkg/daemon/ceph/osd/daemon.go [hunk 2]
```diff
@@ -319,9 +318,26 @@ func getAvailableDevices(context *clusterd.Context, desiredDevices []DesiredDevi
...
+			rawOsds, err := getCephVolumeRawOSDs(context, agent.cluster.Name, agent.cluster.FSID, block, agent.metadataDevice, false)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/daemon.go:326]: Do we need `rawOsds`?

**Comment by satoru-takeuchi** [comment]
> The problem with the new approach is that we have no idea if the device is valid anymore.

How about exporting the target-specific availability fields in Ceph in some way and using this information to judge whether a device is available or not?

https://github.com/ceph/ceph/pull/32634

It's the way @jan--f suggested when I tried to support crypt type devices in the following PR.

https://github.com/ceph/ceph/pull/34375#issuecomment-623301127

File: pkg/daemon/ceph/osd/daemon.go (hunk 1 above)
Comment (travisn) [reply pkg/daemon/ceph/osd/daemon.go:325]: I was doing the same thing as the other place where `getCephVolumeRawOSDs()` is called [here](https://github.com/rook/rook/blob/db12fc86fc0a3ee0b8580d4e362301963aa3dec9/pkg/daemon/ceph/osd/volume.go#L119). Should that be changed as well?

**Comment by travisn** [comment]
Per the comment [here](https://github.com/openshift/ocs-operator/issues/452#issuecomment-623417136), this changed worked.

**Comment by satoru-takeuchi** [comment]
@travisn

> @satoru-takeuchi Yes, that is a side effect. I made that change for consistency between the OSD provisioning and the GetDeviceEmpty() method. Have you perhaps tried installing on a crypt device? If it doesn't work, we should certainly not change it to supported.

Yeah, I found the additional fix is needed to support crypt type. So, the preferable way would be merging this PR first, just to support multipath, and I'll send a PR to support crypt after that.

> @leseb Agreed, we're losing some checks. Better if we had a fix from c-v to allow this scenario. But I think this is safe not to have the additional checks in the PV scenario. The PV should be clean and is only to be consumed by the OSD. We don't have to worry about conflicts of something else using the PV. And if they have a PV that is too small, the OSD will just fail provisioning anyway when c-v tries to create it.

I think so too. However, trying to create OSD even if `ceph-volume inventory` reported the target device is a bit odd. So, how about adopting this approach for now, and changing the implementation after ceph-volume supports mode-specific availability information as described in my last comment. I'll send a PR for this ceph-volume feature to ceph community soon.

**Comment by leseb** [comment]
> > This fix looks like allowing not only "mpath", but also "crypt" (or may also "linear"?). It's because sys.Crypt is already allowed in supportDeviceType() and this type can also bypass ceph-volume inventory check.
> 
> @satoru-takeuchi Yes, that is a side effect. I made that change for consistency between the OSD provisioning and the `GetDeviceEmpty()` method. Have you perhaps tried installing on a crypt device? If it doesn't work, we should certainly not change it to supported.
> 
> > The problem with the new approach is that we have no idea if the device is valid anymore. Although it's somehow "safe" when using PVC in the cloud, when it comes to bare metal I wouldn't be so confident (and it's coming soon). For instance, what if the PV is too small?
> 
> @leseb Agreed, we're losing some checks. Better if we had a fix from c-v to allow this scenario. 

Why not waiting for c-v to fix this? What's so urgent here? A tracker is opened, someone just needs to look into it.

> But I think this is safe not to have the additional checks in the PV scenario. The PV should be clean and is only to be consumed by the OSD. We don't have to worry about conflicts of something else using the PV. And if they have a PV that is too small, the OSD will just fail provisioning anyway when c-v tries to create it.

I think it's ok if we run on the Cloud but definitely not for PV on bare metal.
I think it's a bit too easy to say c-v will fail anyway.
It's really extreme to remove all the checks in place, the proper logging error **just** for one disk type that is definitely not so common.

I would be ok to make an exception (so not this patch) if we were going to backport the fix in a stable branch but that's not the case. This is master code, we just started the 1.4 cycle, so again I don't see why this needs to be done now. If c-v has no fixes once we get close to 1.4 then perhaps we can try something but if @satoru-takeuchi is going to send a PR soon on c-v then let's just wait for it to be merged and backported in Ceph.

Also, we decided that c-v should be our source of truth, if we feel like it's too buggy and we can't get fixes quickly enough then we should probably revisit that decision and implement our own filters.

**Comment by travisn** [comment]
The purpose of calling c-v inventory is only to see if the disk is valid. I don't see the need for this when the OSD is on a PV. With a PV, the OSD is assigned to use exactly this one PV. There is no need to query all disks and see which ones are valid. It's just a single disk. The OSD Is bound to the PV and there really isn't a nice way to recover from a bad PV assignment without manual intervention. So we should just skip the inventory check when on a PV and directly attempt to provision the OSD. If the OSD provisioning fails, then they will anyway have to look at the logs to see why it failed.

**Comment by travisn** [comment]
> @travisn
> 
> > @satoru-takeuchi Yes, that is a side effect. I made that change for consistency between the OSD provisioning and the GetDeviceEmpty() method. Have you perhaps tried installing on a crypt device? If it doesn't work, we should certainly not change it to supported.
> 
> Yeah, I found the additional fix is needed to support crypt type. So, the preferable way would be merging this PR first, just to support multipath, and I'll send a PR to support crypt after that.

Sounds good, i've removed crypt device support and you can add it in the separate PR.

**Comment by satoru-takeuchi** [comment]
FYI.

I'll submit a Ceph issue to use `ceph-volume inventory` again to judge whether a device can be initialized as OSD. I'll also submit a PR to fix this issue soon.

https://tracker.ceph.com/issues/45443

**Comment by travisn** [comment]
For OSDs on PVs, we still have checks in place for a valid device that:
- Does not already have an OSD configured
- Does not have a filesystem or child partition

If we are discovering all devices, the c-v inventory is certainly needed to identify valid devices, but what additional checks will really be needed for PVs? The PV is already bound to the PVC for the OSD, so we should just try to consume it anyway.

**Comment by satoru-takeuchi** [comment]
> what additional checks will really be needed for PVs?

I don't consider that the additional checks are needed, it would works well. However, I consider the current code is a kind of workaround and should eventually be fixed.

The current checks are done in Rook and the corresponding rejected reasons are also generated by Rook.

https://github.com/travisn/rook/blob/5f57b6fd3abbab60b2992171e042598b3fdbf333/pkg/daemon/ceph/osd/daemon.go#L330

However, ideally, it should be done not by Rook, but by Ceph, more precisely, `ceph-volume inventory`. It's straightforward. Although the current "inventory" subcommand doesn't work correctly, we should use it when the problem is fixed. We shouldn't maintain the overlapped logic with Ceph as possible.

**Pull request merged**